extensions/sql-database-projects/src/models/project.ts

```diff
@@ -480,6 +500,54 @@ export class Project {
 		}
 	}
 
+	public addSqlCmdVariableToProjFile(entry: SqlCmdVariableProjectEntry): void {
+		// Remove any entries with the same variable name. It'll be replaced with a new one
+		this.removeFromProjFile(entry);
+
+		const sqlCmdVariableNode = this.projFileXmlDoc.createElement(constants.SqlCmdVariable);
+		sqlCmdVariableNode.setAttribute(constants.Include, entry.variableName);
+		this.addSqlCmdVariableChildren(sqlCmdVariableNode, entry);
+		this.findOrCreateItemGroup(constants.SqlCmdVariable).appendChild(sqlCmdVariableNode);
+
+		// add to the project's loaded sqlcmd variables
+		this.sqlCmdVariables[entry.variableName] = <string>entry.defaultValue;
+	}
+
+	private addSqlCmdVariableChildren(sqlCmdVariableNode: any, entry: SqlCmdVariableProjectEntry): void {
+		// add default value
+		const defaultValueNode = this.projFileXmlDoc.createElement(constants.DefaultValue);
+		const defaultValueText = this.projFileXmlDoc.createTextNode(entry.defaultValue);
+		defaultValueNode.appendChild(defaultValueText);
+		sqlCmdVariableNode.appendChild(defaultValueNode);
+
+		// add value node which is in the format $(SqlCmdVar__x)
+		const valueNode = this.projFileXmlDoc.createElement(constants.Value);
+		const valueText = this.projFileXmlDoc.createTextNode(`$(SqlCmdVar__${this.getNextSqlCmdVariableCounter()})`);
+		valueNode.appendChild(valueText);
+		sqlCmdVariableNode.appendChild(valueNode);
+	}
+
+	/**
+	 * returns the next number that should be used for the new SqlCmd Variable. Old numbers don't get reused even if a SqlCmd Variable
+	 * gets removed from the project
+	 */
+	private getNextSqlCmdVariableCounter(): number {
+		const sqlCmdVariableNodes = this.projFileXmlDoc.documentElement.getElementsByTagName(constants.SqlCmdVariable);
+		let highestNumber = 1;
+
+		for (let i = 0; i < sqlCmdVariableNodes.length; i++) {
```

Comment:
What does this counter do?  Like, I see it in the sqlproj file, but I've always seen sqlcmdvars consumed by their name in the `Include` attribute rather than this.

I ask because I'm wondering whether this should be something that's determined/saved as part of the SqlCmdVariableProjectEntry rather than calculated off of sqlproj file contents...

Reply:
the counter is used in SSDT to map the sqlcmdvar to local value stored in the .user file. It's not something that's currently supported in our ADS sql database projects. 

.user file:
```
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <SqlCmdVar__4>col</SqlCmdVar__4>
    <SqlCmdVar__9>localval</SqlCmdVar__9>
  </PropertyGroup>
</Project>
```

sqlcmdvars in .sqlproj:
```
  <ItemGroup>
    <SqlCmdVariable Include="columnset">
      <DefaultValue>columnset-2020-4-28-12-22</DefaultValue>
      <Value>$(SqlCmdVar__4)</Value>
    </SqlCmdVariable>
    <SqlCmdVariable Include="testdb">
      <DefaultValue>testdb-2020-5-27-10-40</DefaultValue>
      <Value>$(SqlCmdVar__6)</Value>
    </SqlCmdVariable>
    <SqlCmdVariable Include="testvar">
      <DefaultValue>defaultVal</DefaultValue>
      <Value>$(SqlCmdVar__9)</Value>
    </SqlCmdVariable>
  </ItemGroup>
```